### PR TITLE
add 'params' method for parametrizing tests

### PR DESCRIPTION
**Pull request by dimastark**

### CODE REVIEW

**Comment by silkentrance** [comment]
@dimastark the failing tests are unrelated, just wanted to make sure that your PR did not hide them.

**Comment by silkentrance** [comment]
@dimastark please rebase to master, #74 has been fixed.

**Comment by pana-cc** [comment]
I would agree with @silkentrance I too think the implementation should be close to what is discussed in #60 

As for the public API I don't have strong opinion, the multiple @params seems interesting, I just hope it is scalable enough, a single @params with an array.

**Comment by dimastark** [comment]
I think multiple `@params` are better, because you can add `skip` for some of them, for example.

I'll fix pr a little later.

**Comment by pana-cc** [comment]
@dimastark awesome, that way we should be able to also use @params.only(...) to run and debug a single test.

File: package.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
   ],
   "license": "Apache-2.0",
   "dependencies": {
-    "@types/mocha": "*",
+    "@types/mocha": "5.2.0",
```
Comment (dimastark) [inline package.json]: Something broken in "@types/mocha" with version "5.2.1".

Comment (dimastark) [reply package.json]:
```typescript
node_modules/@types/mocha/index.d.ts:456:11 - error TS2420: Class 'Suite' incorrectly implements interface 'ISuite'.
  Type 'Suite' provides no match for the signature '<TFunction extends Function>(target: TFunction): void | TFunction'.
```
and
```typescript
node_modules/@types/mocha/index.d.ts:499:11 - error TS2420: Class 'Test' incorrectly implements interface 'ITest'.
  Type 'Test' provides no match for the signature '<T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>): void | TypedPropertyDescriptor<T>'.
```

**Comment by dimastark** [comment]
It's ready for review

File: index.ts [hunk 2]
```diff
@@ -200,21 +204,50 @@ function suiteClassCallback(target: SuiteCtor, context: TestFunctions) {
...
+						shouldOnly = mark === onlySymbol;
+						shouldPending = mark === pendingSymbol;
+
+						if (shouldPending && !shouldSkip && !shouldOnly) {
+							context.it.skip(testName);
+						}
+
+						applyTestFunc(testFunc, parametersTestName, method, [parameter], method.length <= 1);
+                    });
+                } else {
```
Comment (pana-cc) [inline index.ts]: Maybe white-space is skewed here. I hate to return PRs for such non-productive stuff so I'd be ok to merge it. But we should add tslint and autofix scripts in future. For reference: https://github.com/NativeScript/css-parser/blob/1f727144b8b6dfe178f63f4876183bb67976641a/package.json#L10

File: test.ts [hunk 3]
```diff
@@ -75,124 +75,44 @@ function trimEmptyLines(str: string, eliminateAll = false): string {
...
+    @params({ target: "es5", ts: "only.suite" })
+    @params({ target: "es6", ts: "only.suite" })
+    @params({ target: "es5", ts: "pending.suite" })
+    @params({ target: "es6", ts: "pending.suite" })
+    @params({ target: "es5", ts: "retries.suite" })
+    @params({ target: "es6", ts: "retries.suite" })
+    @params({ target: "es6", ts: "context.suite" })
+    @params({ target: "es5", ts: "abstract.inheritance.suite" })
+    @params({ target: "es6", ts: "abstract.inheritance.suite" })
+    @params({ target: "es5", ts: "suite.inheritance.suite" })
```
Comment (pana-cc) [inline test.ts]: <3

**Comment by silkentrance** [comment]
The fix for `@types/mocha` is no longer required. In PR #81 the problem has been resolved.

**Comment by silkentrance** [comment]
@dimastark Some of the reviews have been outdated by the last commit, please take them into account.

**Comment by silkentrance** [comment]
@pana-cc do async tests still work, do we have any tests that would fail?

**Comment by pana-cc** [comment]
Rebased along with adding tslint in #82 
Thank you for the contribution.
Please follow the discussion there.

**Comment by pana-cc** [comment]
@silkentrance 
Do you refer these tests?
https://github.com/pana-cc/mocha-typescript/blob/master/tests/ts/test.v2.suite.ts#L30
https://github.com/pana-cc/mocha-typescript/blob/61835f22368ec28920a49b989d98db28dcd4648e/tests/ts/test.v2.suite.ts#L121
https://github.com/pana-cc/mocha-typescript/blob/61835f22368ec28920a49b989d98db28dcd4648e/tests/ts/test.v2.suite.ts#L205

**Comment by silkentrance** [comment]
@dimastark please enable the Package tests again

**Comment by silkentrance** [comment]
@pana-cc yes, just found them 👍 wonder why these still work, though.

**Comment by pana-cc** [comment]
@silkentrance I've made a separate branch "params" let's take it from here, could you enable the Package tests in that branch?

**Pull request closed without merging**